// nuxt-viewport/lib/index.js
'use strict';

require('js-cookie');

const DEFAULT_OPTIONS = {
  breakpoints: {
    desktop: 1024,
    desktopMedium: 1280,
    desktopWide: 1600,
    mobile: 320,
    mobileMedium: 375,
    mobileWide: 425,
    tablet: 768
  },
  cookieName: "viewport",
  defaultBreakpoints: {
    desktop: "desktop",
    mobile: "mobile",
    tablet: "tablet"
  },
  fallbackBreakpoint: "desktop"
};

const nuxtModule = function(customOptions) {
  const {nuxt} = this;
  const options = {
    ...DEFAULT_OPTIONS,
    ...this.options.viewport || customOptions || {}
  };
  this.addPlugin({
    filename: "viewport.client.js",
    options,
    src: require.resolve("./runtime/plugin.client")
  });
  this.addPlugin({
    filename: "viewport.server.js",
    options,
    src: require.resolve("./runtime/plugin.server")
  });
  nuxt.options.alias["~viewport"] = require.resolve("./runtime/viewport");
  nuxt.options.build.transpile.push(__dirname, "nuxt-viewport");
};
nuxtModule.meta = require("../package.json");

module.exports = nuxtModule;
